src/audio/module_adapter/module_adapter.c

```diff
@@ -869,144 +790,6 @@ static int module_adapter_audio_stream_copy_1to1(struct comp_dev *dev)
 	return ret;
 }
 
-static int module_adapter_audio_stream_type_copy(struct comp_dev *dev)
-{
-	struct comp_buffer __sparse_cache *source_c[PLATFORM_MAX_STREAMS];
-	struct comp_buffer __sparse_cache *sinks_c[PLATFORM_MAX_STREAMS];
-	struct processing_module *mod = comp_get_drvdata(dev);
-	struct list_item *blist;
-	uint32_t num_input_buffers, num_output_buffers;
-	int ret, i = 0;
-
-	/* handle special case of HOST/DAI type components */
-	if (dev->ipc_config.type == SOF_COMP_HOST || dev->ipc_config.type == SOF_COMP_DAI)
-#if CONFIG_IPC_MAJOR_3
-		return module_process_legacy(mod, NULL, 0, NULL, 0);
-#else
-		return module_process_stream(mod, NULL, 0, NULL, 0);
-#endif
```

Comment:
just because of this 1-line difference the whole function is duplicated?

Reply:
yes, this is deserved, since process is time critical, code looks duplicated, however, it does not, due to ipc3 and ipc4 build separately.

Reply:
@btian1 The code is is duplicated in the codebase and if we find a bug in this long function, it now needs to be fixed in two places.

But right, it is a compromise to make, we don't want any added runtime cost in functions like copy/process.

I'm looking at @ranj063 's 5f6f22d25e3a74eca84ca01ba79949d3ebbce960 and your commit 4cd928007edf10ed5ac6fa4022a2eb0a348e651a . So this is a anyways a special case for DAI.

Can we call "module_process_for_dai()" in generic.c and then have a ifdef in header to call (inline) either legacy or stream (IPC3 or IPC4)? Then we can share this function as-is and have no runtime penalty.

Reply:
I still don't think we should optimise even the hot path to that level. At least I don't think we're currently at that stage. As I said again - I do think, that we have to optimise to the maximum internal processing loops. There - yes, every instruction, every cycle counts. But anything before or after that loop - I don't think we're saving function calls there. Those paths should avoid wasting cycles meaninglessly of course, but I don't think we're getting rid of all function calls at that level.

Reply:
> Can we call "module_process_for_dai()" in generic.c

@kv2019i maybe "edge" since it is for both DAI and host?

Reply:
> module_process_legacy was called in multiple place, so change name is not suitable.

@btian1 I think what @kv2019i meant was
```
static inline int module_process_for_dai(...)
{
#if CONFIG_IPC_MAJOR_3
    return module_process_legacy(....);
#else
    return module_process_stream(...);
#endif
}
```
that way you remove the `#if` from the function body, incur no run-time overhead, and keep both functions to be called elsewhere as needed.

> also as this is the common file for ipc3 and ipc4, expectation is zero #if IPC3/IPC4.

I'm not aware of this expectation. We're certainly rather far from that. I wouldn't be against it - I'd be happy to have zero `#if`s in .c code altogether, like the Linux kernel does, but SOF is more tolerant to that. So, in general, getting rid of them is good, but not at any cost, I would say.